### PR TITLE
Refactor `core:text/table`

### LINKED ISSUE #3432: text/table uses bytes instead of runes for length
## Context
The `text/table` library is using `len` where `utf8.rune_count` would probably be more appropriate. See the following program and its output.

```odin
package main

import "core:os"
import "core:text/table"

main :: proc() {
    t: table.Table
    table.init(&t)
    table.caption(&t, "Tɛst")
    table.row(&t, "Hellope!")
    table.row(&t, "Hɛllope!")

    stdout := os.stream_from_handle(os.stdout)
    table.write_ascii_table(stdout, &t)
}
```

The `ɛ` is a two-byte UTF-8 rune here.

## Result
```
+---------+
|  Tɛst  |
+---------+
|Hellope! |
|Hɛllope!|
+---------+
```

## Expected
```
+--------+
|  Tɛst  |
+--------+
|Hellope!|
|Hɛllope!|
+--------+
```

### PR DESCRIPTION
**Pull request by Feoramund**
_Several kilobytes of lookup tables and one grapheme segmentation algorithm later..._

unicode-table-alignment

Resolves #3432

### CODE REVIEW

**Comment by oskarnp** [comment]
This is awesome. However, I am just not a fan of breaking the API unless it is neccessary?

**Comment by oskarnp** [comment]
I'll review next week.

**Comment by Feoramund** [comment]
I changed `colw` to make calculations easier, `set_cell_value_and_alignment` because it was inconsistent with `set_cell_value` and only took a string (not an `any`), `build` to be the place where `Width_Proc` is specified (though this could have a default value if needed), renamed the proc from `ascii` to `plain` per the discussion in the original issue, and `write_text_align` to take a `space` argument instead of `colw`.

The arguments were rotated on `write_text_align` to purposefully break the API because the semantics of the arguments changed but the types did not.

File: core/text/table/table.odin [hunk 1]
```diff
@@ -37,10 +43,19 @@ Table :: struct {
...
 
 	// The following are computed on build()
-	colw: [dynamic]int, // Width of each column (including padding, excluding borders)
+	colw: [dynamic]int, // Width of each column (excluding padding and borders)
 	tblw: int,          // Width of entire table (including padding, excluding borders)
 }
 
+ascii_width_proc :: proc(str: string) -> int {
+	return len(str)
+}
```
Comment (oskarnp) [inline core/text/table/table.odin:58]: Is this really needed, considering ascii is a subset of utf-8?

File: core/text/table/table.odin [hunk 2]
```diff
@@ -165,7 +180,7 @@ first_row :: proc(tbl: ^Table) -> int {
 	return header_row(tbl)+1 if tbl.has_header_row else 0
 }
 
-build :: proc(tbl: ^Table) {
+build :: proc(tbl: ^Table, width_proc: Width_Proc) {
```
Comment (oskarnp) [inline core/text/table/table.odin:290]: I think I would prefer a default Width_Proc here. Default to `unicode_width_proc`.

But considering previous comment about ascii being a subset, is Width_Proc needed at all? Any use case for custom width procs you can think of?

File: core/text/table/table.odin [hunk 3]
```diff
@@ -251,8 +265,10 @@ write_html_table :: proc(w: io.Writer, tbl: ^Table) {
 }
 
 write_plain_table :: proc(w: io.Writer, tbl: ^Table) {
+	WIDTH_PROC :: unicode_width_proc
+
 	if tbl.dirty {
-		build(tbl)
+		build(tbl, WIDTH_PROC)
```
Comment (oskarnp) [inline core/text/table/table.odin:378]: If we are to support custom Width_Proc's they should be stored on the `Table` struct at the very least to avoid this situation. Or just remove custom Width_Proc's. Again, is there a use case?

File: core/text/table/table.odin [hunk 4]
```diff
@@ -152,22 +164,116 @@ header :: proc(tbl: ^Table, values: ..any, loc := #caller_location) {
...
 			panic("Cannot create empty row unless the number of columns is known in advance")
 		} else {
 			tbl.nr_cols = len(values)
 		}
 	}
+
 	tbl.nr_rows += 1
-	for col in 0..<tbl.nr_cols {
+
+	for col in 0 ..< tbl.nr_cols {
```
Comment (oskarnp) [inline core/text/table/table.odin]: `a..<b` (without surrounding space) is the preferred Odin style.

File: core/text/table/table.odin [hunk 5]
```diff
@@ -152,22 +164,116 @@ header :: proc(tbl: ^Table, values: ..any, loc := #caller_location) {
...
+		if len(aligned_values) == 0 {
+			panic("Cannot create empty row unless the number of columns is known in advance")
+		} else {
+			tbl.nr_cols = len(aligned_values)
+		}
 	}
+
+	tbl.nr_rows += 1
+
+	for col in 0 ..< tbl.nr_cols {
```
Comment (oskarnp) [inline core/text/table/table.odin:258]: `a..<b` (without surrounding space) is the preferred Odin style.

File: core/text/table/table.odin [hunk 6]
```diff
@@ -152,22 +164,116 @@ header :: proc(tbl: ^Table, values: ..any, loc := #caller_location) {
...
+		if len(values) == 0 {
+			panic("Cannot create empty row unless the number of columns is known in advance")
+		} else {
+			tbl.nr_cols = len(values)
+		}
+	}
+
+	tbl.nr_rows += 1
+
+	for col in 0 ..< tbl.nr_cols {
```
Comment (oskarnp) [inline core/text/table/table.odin:220]: `a..<b` (without surrounding space) is the preferred Odin style.

File: core/text/table/table.odin [hunk 7]
```diff
@@ -37,10 +49,19 @@ Table :: struct {
...
+	colw: [dynamic]int, // Width of each column (excluding padding and borders)
 	tblw: int,          // Width of entire table (including padding, excluding borders)
 }
 
+ascii_width_proc :: proc(str: string) -> int {
+	return len(str)
+}
+
+unicode_width_proc :: proc(str: string) -> (width: int) {
+	_, _, width = #force_inline utf8.grapheme_count(str)
```
Comment (oskarnp) [inline core/text/table/table.odin:61]: `utf8.grapheme_count()` calls `utf8.decode_grapheme_clusters()` which allocates implicitly using `context.allocator`. Isn't this a hidden allocation and a memory leak?
All allocators used should be specified up front at `init`.

File: core/text/table/table.odin [hunk 8]
```diff
@@ -137,22 +165,116 @@ header :: proc(tbl: ^Table, values: ..any, loc := #caller_location) {
...
+		if len(values) == 0 {
+			panic("Cannot create empty row unless the number of columns is known in advance")
+		} else {
+			tbl.nr_cols = len(values)
+		}
+	}
+
+	tbl.nr_rows += 1
+
+	for col in 0 ..< tbl.nr_cols {
```
Comment (oskarnp) [inline core/text/table/table.odin:239]: `a..<b` (without surrounding space) is the preferred Odin style.

File: core/text/table/table.odin [hunk 9]
```diff
@@ -306,45 +425,58 @@ write_ascii_table :: proc(w: io.Writer, tbl: ^Table) {
...
-					io.write_byte(w, ':')
-				}
+	start := 0
+
+	if tbl.has_header_row {
+		row := header_row(tbl)
+
+		write_row(w, tbl, row, .Center)
+
+		for col in 0 ..< tbl.nr_cols {
```
Comment (oskarnp) [inline core/text/table/table.odin:453]: `a..<b` (without surrounding space) is the preferred Odin style.

I'll stop pointing these out now. Please fix them all :)

**Review by oskarnp — CHANGES_REQUESTED** [review]

File: core/text/table/table.odin (hunk 3 above)
Comment (Kelimion) [reply core/text/table/table.odin:378]: I suppose that if you have a specific font in mind, you could parse its kerning table and glyph info to be more specific. But are you still working in plain text using this package or doing tables in a graphical context?

File: core/text/table/table.odin (hunk 1 above)
Comment (Feoramund) [reply core/text/table/table.odin:58]: Yes. `grapheme_count` has to iterate by UTF-8 codepoints then address several lookup tables by binary search. It's not fast at all. If someone knows in advance that all the text they're handling is ASCII, then it's better to use this proc rather than the Unicode version.

File: core/text/table/table.odin (hunk 3 above)
Comment (Feoramund) [reply core/text/table/table.odin:378]: I did consider if the proc should be on the table or not. I concluded that it shouldn't be on the table, because how the table is printed is separate from what data is stored on the table and HTML tables do not measure width.

File: core/text/table/table.odin (hunk 2 above)
Comment (oskarnp) [reply core/text/table/table.odin:290]: Given your comment about `grapheme_count` I change my mind here. Leave it as it is.

File: core/text/table/table.odin (hunk 3 above)
Comment (oskarnp) [reply core/text/table/table.odin:378]: Maybe we should consider `if tbl.dirty` at that point to be an error instead to avoid this. 🤔

File: core/text/table/table.odin (hunk 4 above)
Comment (Feoramund) [reply core/text/table/table.odin]: I find this style `a ..< b` to be more readable and tend to use it. I did a search to confirm your statement. At the very least, it's more common (~450 instances of `a..<b` versus ~50 of `a ..< b`). I can change it if it's an issue.

File: core/text/table/table.odin (hunk 7 above)
Comment (Feoramund) [reply core/text/table/table.odin:61]:
```
Inputs:
- str: The input string.
- track_graphemes: Whether or not to allocate and return `graphemes` with extra data about each grapheme.
- allocator: (default: context.allocator)

```

```
grapheme_count :: proc(str: string) -> (graphemes, runes, width: int) {
	_, graphemes, runes, width = decode_grapheme_clusters(str, false)
```

```
				if track_graphemes {
					append(&graphemes, Grapheme{
						byte_index,
						rune_count,
						width - last_width,
					})
				}
```

This interpretation is incorrect. In addition, `grapheme_count` is tested under the test suite, and if a memory leak occurs, it would show up in the warnings.

Comment (oskarnp) [reply core/text/table/table.odin:61]: I stand corrected. Ignore my comment.

**Review by oskarnp — CHANGES_REQUESTED** [review]
I think `doc.odin` needs some more love. Could you include the example from your PR? And also showcase using different Width_Proc's.

File: core/text/table/table.odin (hunk 3 above)
Comment (Feoramund) [reply core/text/table/table.odin:378]: Do you want `tbl.dirty` done away with?

Comment (oskarnp) [reply core/text/table/table.odin:378]: What would you suggest?

Comment (Feoramund) [reply core/text/table/table.odin:378]: It's a caching situation. The advantage of having `tbl.dirty` around is that it allows re-printing of the table without needing to recalculate column widths. This is assuming there's a situation where a table has been built and is not often changed, but is shown on more than one or two occasions. However, I think if someone wants to really save time, they're better off writing the table once and storing that as a string and avoiding all of the calculations altogether.

In summary, `dirty` avoids recalculating column widths, but in the end, you're still doing all the work to do the printing, which is the bulk of the work being done (all the calls to `io`'s API which potentially has to branch with each call).

My suggestion is to remove `dirty` and offer string caching as a way to save time, if that's a goal for the programmer.

Comment (oskarnp) [reply core/text/table/table.odin:378]: If you don't mind making those changes, then yes I agree that would be a better solution and we should get rid of it.

Comment (oskarnp) [reply core/text/table/table.odin:378]: The user could easily do the string caching themselves passing a custom `io.Writer` so I think we can skip that and just remove `dirty`.

**Comment by Feoramund** [comment]
All set.

**Comment by oskarnp** [comment]
I can't get the docs.odin to render that unicode table properly. Both GitHub and Sublime Text shows it incorrectly.
Did you paste that properly? Or are they both wrong.

**Comment by Feoramund** [comment]
I have confirmed that Alacritty, kitty, WezTerm (with Thai out of alignment for some reason but otherwise fine), Konsole, QTerminal (same Thai issue), LXTerminal, and MATE terminal render it properly. I may not have good all-around Thai fonts installed, but it's otherwise fine on my end.

GitHub's display is browser-dependent, and it certainly doesn't render properly on my end, but I wouldn't expect it to.

I found a Sublime Text editor package for my distribution of Linux and installed it to have a look. With the default settings, all of the Unicode lines are out of alignment (except the actual Younger Futhark runes, amusingly enough). I'm able to get some of them to align if I set certain monospace fonts (Ubuntu Mono), but as this is the first time I've ever used Sublime Text before, I am uncertain if there may be some setting I'd have to toggle to get it to work properly.

It's very strange that it's getting CJK wrong, as they're the most easily predictable in terms of width. Even with a CJK font set, it's still not aligned. It also has issues with rendering the excessive diacritics line, as in, glyph placeholders are appearing.

I tried manually inserting spaces to see if I could align the Chinese row, and even with extra spaces, it's mis-aligned. I.e. there's no amount of space characters I can manually put in that would make the `|` line up. I'm inclined to say this is a rendering issue.

**Comment by oskarnp** [comment]
@gingerBill @Kelimion LGTM

Great work, @Feoramund

**Comment by Kelimion** [comment]
Thanks as always, @Feoramund.
